📐 Add a basic edit layout

pdrj opened the pull request:
## Info

### Summary
This adds a basic edit layout

### Technical considerations
<img width="132" alt="Screenshot 2020-04-23 at 14 34 02" src="https://user-images.githubusercontent.com/53510080/80099895-7cce9980-856f-11ea-8fd4-d11f0000c778.png">

- I had to wrap the bottom line of ☝️ in linear layout because I couldn't [in a reasonable amount of time] make all 3 elements to be centered related to the header with just a flat constraint layout 😢If you know a way to do it, please share :) 
- The second commit wraps the whole thing [except the close button, handle and time indicator] in a nested scroll view - otherwise the wheel placeholder gets cut off. I don't like it, but I don't think there's a way around that. 
- We have `AlphaSlideAction` in the fragment. I don't think it worked for me before (or I can't see it), but now it just can't work as it's referencing a `Group` and setting group's alpha has no effect. **Should we just remove the animation? Can you see it working on your device?** 
- I have not handled changing the labels on start/end time labels in case the TE has no duration or is not running yet. For this to be implemented, I'd have to have my changes regarding displaying the correct time on the indicator. I could rebase on that branch and add those features, I just don't know what the expectation on this issue is, since it's just an `S`.

### Why is it done like this and not some other way?
I'm open to suggestions.

### Relationships

Closes #202 

## Repository health

### Pull Request checklist
<!-- MANDATORY: Before submiting your Pull Request, make sure that -->
<!-- all the items below are verified and the boxes are checked -->
- [x] The code does not hinder the app's performance
- [x] The code does not introduce regressions
- [x] The code in this PR is tested thoroughly enough that changing the behavior to an incorrect one breaks tests

## Reviewing

### Pointers
Use this with keyboard on/off; expand, collapse. See if it behaves correctly.

### Review template
<!-- MANDATORY: This is what reviewers should consider when reviewing your Pull request -->
<!-- Feel free to remove items that UI related if your code does not touch UI -->

<!-- General -->
- [ ] The code is thoroughly tested in well written and resilient manner
- [ ] The code is performant and leads to an optimal user experience
- [ ] The code is testable and does not cut corners for the sake of delivering
- [ ] The code considers older platforms well and will work in older devices just like it will in new ones
<!-- UI related -->
- [ ] The code considers light and dark theme
- [ ] The code considers RTL and does not mention left/right
- [ ] The code considers i18n
  - [ ] There are no hardcoded strings in the UI
  - [ ] All strings are included in every language we support OR there's an issue requesting new translations
- [ ] There are no hard-coded numbers referring to dimensions or font sizes
- [ ] The code considers UI elements that are similar and extracts their similarities into styles to avoid duplication

## Merge permissions
Anyone

semanticer commented on timer/src/main/res/layout/fragment_dialog_start_edit.xml:
```diff
@@ -36,78 +36,187 @@
+                        style="@style/Widget.MaterialComponents.Chip.Action"
+                        android:layout_width="wrap_content"
+                        android:layout_height="wrap_content"
+                        android:text="Add a Project"
```
Maybe it's time to extract these strings? Not sure what the plan is @heytherewill ?

pdrj replied:
Yeah... For sure that needs to be done at some point. Extracting is not much work, so I'll just do it.

daividssilverio commented:
> I have not handled changing the labels on start/end time labels in case the TE has no duration or is not running yet. For this to be implemented, I'd have to have my changes regarding displaying the correct time on the indicator. I could rebase on that branch and add those features, I just don't know what the expectation on this issue is, since it's just an S.

Oh, the idea was to have that done in this PR, but since I didn't notice that dependency, I think we can be flexible with that.
I can create another issue to handle the stop timer as an `size-XS` that can be handled after #200 is merged.

➿ @heytherewill

daividssilverio commented on timer/src/main/res/layout/fragment_dialog_start_edit.xml:
```diff
@@ -36,78 +36,187 @@
 
-    <com.google.android.material.textfield.TextInputLayout
-        android:id="@+id/time_entry_description_input_layout"
+    <androidx.core.widget.NestedScrollView
```
We can simplify a bit this layout by making `NestedScrollView` its root and removing the extra nested `ConstraintLayout`, what do you think?

pdrj replied:
Not really, then the 'x' button and the handle would also scroll and we don't want that. Am I missing something @daividssilverio?

daividssilverio replied:
Oh, I didn't notice that. Damn, forget this request then.

daividssilverio commented:
> I had to wrap the bottom line of ☝️ in linear layout because I couldn't [in a reasonable amount of time] make all 3 elements to be centered related to the header with just a flat constraint layout 😢If you know a way to do it, please share :)

As a matter of fact we managed to do it in flat way on the old app. It would look like something like this:

https://github.com/toggl/mobileapp/blob/2fba1cdea3fa55f107f331d83972f3a677e5509b/Toggl.Droid/Resources/layout/EditDurationActivity.axml#L22-L181

Basically, the idea is:
- to have a vertical guideline that divides the screen in two
- each header (start | stop) is centered on its side of the divider (start or end)
- each [`date text` <-> `round icon` <-> `start time`] chain is centered on its side of the divider
- the `stop` label is centered likewise on its side of the divider

It works quite alright.

pdrj commented:
@daividssilverio @semanticer 
I've re-enabled the animation - I didn't notice it before because it's quite subtle and you have to expand/collapse slowly. It works now and is a nice touch 👍 

I've also removed the nested linear layout as suggested by Daivid.

I've also handled the visibility and state of start/stop labels in different scenarios. This is quite messy TBH, if you've got better ideas - I'm open to suggestions. The commit is https://github.com/toggl/android/pull/223/commits/881507d3016521b1752a57c679a275d235b6474f.

I've added handling for the billable chip too - it now dispatches the action and is checked/unchecked according to state.

pdrj commented:
Oh, and I've merged my unmerged PR into this so I can use the new fields of EditableTimeEntry.

heytherewill commented:
I didn't review, but a bunch of string were added; don't forget to add the translations for the languages we support as well :)

pdrj commented:
I'll add these that can be exported form the old app @heytherewill. There's also one that I came up with as a sort of placeholder - I asked in the unification project channel how the view should actually look in the state that the added string is describing.

pdrj commented:
@heytherewill 
We're missing:
- Add a project
- End
- Time entry not started yet [potentially, this one probably won't be used anyway]

Should I create 'translation issues' for the first two?

daividssilverio commented:
@pdrj 
> I've also handled the visibility and state of start/stop labels in different scenarios. This is quite messy TBH, if you've got better ideas - I'm open to suggestions. The commit is 881507d.

Stop label handling is all good, but what about those "time entry not started yet"? That is not how we handle those cases, right? At least the start time should be set at any point 🤔 even if defaulting to "now".

Although I think that's fine for now, we need to go through designs and specs to decide if we are gonna keep that. ping @heytherewill

pdrj commented:
Related thread https://toggl.slack.com/archives/CS25QL11P/p1587728792000600

heytherewill commented on common-android/src/main/java/com/toggl/common/ViewExtensions.kt:
```diff
@@ -0,0 +1,13 @@
+import android.view.MotionEvent
+import android.view.View
+
+fun View.addInterceptingOnClickListener(action: () -> Unit) {
+    this.setOnTouchListener { _, event ->
+        if (event.action == MotionEvent.ACTION_UP) {
+            action()
+            true
+        } else false
+    }
+}
```
Why do we need this to only happen in touch up? 🤔

pdrj replied:
I usually use this one to mean 'a click'. I know it's not exactly the same as click though :) - would you like me to track how long the press took, what the coordinates were, etc?

heytherewill commented on common-android/src/main/res/values-fr/strings.xml:
```diff
@@ -13,6 +13,10 @@
+    <string name="add_tags">Ajoutez les labels</string>
+    <string name="billable">Facturable</string>
+    <string name="start">Début</string>
+    <string name="stop_timer">Stop timer</string>
```
This is in english

pdrj replied:
These English strings are gone in the latest commit

heytherewill commented on common-android/src/main/res/values-pt/strings.xml:
```diff
@@ -13,6 +13,11 @@
     <string name="continue_this">Continuar</string>
     <string name="delete">Excluir</string>
     <string name="undo">Desfazer</string>
+    <string name="add_project">Add a project</string>
```
This is english

heytherewill commented on common-android/src/main/res/values-ja/strings.xml:
```diff
@@ -13,6 +13,10 @@
+    <string name="add_tags">タグを追加</string>
+    <string name="billable">請求可能</string>
+    <string name="start">開始</string>
+    <string name="stop_timer">Stop timer</string>
```
English

heytherewill commented on common-android/src/main/res/values/colors.xml:
```diff
@@ -19,4 +19,6 @@
   <color name="button_active">#06AAF5</color>
   <color name="button_active_background">#2606AAF5</color>
   <color name="button_inactive">#B5BCC0</color>
+  <color name="clickable_text_primary">#328FFF</color>
+  <color name="clickable_text_secondary">#2196F3</color>
```
Primary and secondary are enabled and disabled? If so, we should consider renaming it

pdrj replied:
Nope, the primary is the time label, the secondary is the date label. I named it like so because the time one seems 'bolder' in the designs.
<img width="182" alt="Screenshot 2020-04-27 at 14 07 54" src="https://user-images.githubusercontent.com/53510080/80370464-7fdbc980-8890-11ea-99ec-01b998ee22c7.png">

heytherewill commented on timer/src/main/java/com/toggl/timer/running/domain/RunningTimeEntryReducer.kt:
```diff
@@ -31,7 +33,7 @@ class RunningTimeEntryReducer @Inject constructor(
                     val entryToOpen = timeEntries.runningTimeEntryOrNull()
                         ?.run(EditableTimeEntry.Companion::fromSingle)
-                        ?: EditableTimeEntry.empty(defaultWorkspaceId())
+                        ?: EditableTimeEntry.withStartTime(defaultWorkspaceId(), timeService.now())
```
Wait a second, this is not correct. The start time should not be set when we open the card, but rather when the entry is started. I understand that you need the `now` time for displaying, but this is not the correct way of doing it because then the user might get unexpected start times.

pdrj replied:
I thought that was the decision from https://toggl.slack.com/archives/CS25QL11P/p1587728792000600 Otherwise, we'd have to update the start label each minute?

daividssilverio replied:
Yeah... I think we can start the `EditableTimeEntry` with a null `startTime` and handle displaying "now" from the UI side until the user interacts with the wheel or one of the time pickers. What do you think @heytherewill @pdrj ?

pdrj replied:
Yep that's how it's done now @daividssilverio [we're obviously missing the editing part].

pdrj commented:
Scrolling issue: #236 
Translations: #235

daividssilverio commented on timer/src/main/java/com/toggl/timer/startedit/ui/StartEditDialogFragment.kt:
```diff
@@ -104,7 +109,7 @@ class StartEditDialogFragment : BottomSheetDialogFragment() {
             store.state
                 .filter { it.editableTimeEntry != null }
-                .map { BottomControlPanelParams.fromState(it) }
+                .map { BottomControlPanelParams(it.editableTimeEntry!!, it.isEditableInProWorkspace()) }
```
I wish the compiler could be smarter here... those `!!` make me anxious.

heytherewill replied:
Well, what we should be doing here is throwing the appropriate exception and letting the compiler infer that it can't be null 🤔

pdrj replied:
You mean in `map` @heytherewill ? Like 
```
                .map {
                    it.editableTimeEntry ?: throw EditableTimeEntryShouldNotBeNullException()
                    BottomControlPanelParams(it.editableTimeEntry, it.isEditableInProWorkspace()) 
                }
```

pdrj replied:
That would let us omit the !!, but I think @daividssilverio means that the compiler should already know that it can't be null, thanks to the filter.

pdrj commented:
I agree @daividssilverio, applied the suggestion.
